### PR TITLE
Implemented lingering potions and area effect clouds

### PR DESCRIPTION
**Pull request by JustTalDevelops**

### CODE REVIEW

**Review by DaPigGuy — APPROVED** [review]

File: server/entity/splashable.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			}
+
+			splashed := otherE.(Living)
+			for _, eff := range effects {
+				if p, ok := eff.Type().(effect.PotentType); ok {
+					splashed.AddEffect(effect.NewInstant(p.WithPotency(f), eff.Level()))
+					continue
+				}
+
+				dur := time.Duration(float64(eff.Duration()) * 0.75 * f)
```
Comment (DaPigGuy) [inline server/entity/splashable.go:57]: For lingering potions, rather than multiplying by `0.75`, the duration should be multiplied by `0.25`

Comment (JustTalDevelops) [reply server/entity/splashable.go]: Updated.

File: server/entity/area_effect_cloud.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+// BBox ...
+func (a *AreaEffectCloud) BBox() cube.BBox {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return cube.Box(-a.radius, 0, -a.radius, a.radius, 0.5, a.radius)
+}
+
+// Tick ...
+func (a *AreaEffectCloud) Tick(w *world.World, _ int64) {
```
Comment (T14Raptor) [inline server/entity/area_effect_cloud.go:96]: Really don't like how this method can lock/unlock at least 3 times, plus 2*the amount of entities in radius

File: server/entity/splashable.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+				}
+				splashed.AddEffect(effect.New(eff.Type().(effect.LastingType), eff.Level(), dur))
+			}
+		}
+	} else if s.t == potion.Water() {
+		switch result := res.(type) {
+		case trace.BlockResult:
+			blockPos := result.BlockPosition().Side(result.Face())
+			if w.Block(blockPos) == fire() {
+				w.SetBlock(blockPos, air(), nil)
```
Comment (T14Raptor) [inline server/entity/splashable.go:75]: Can just use nil here

File: server/entity/splashable.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			}
+
+			for _, f := range cube.HorizontalFaces() {
+				if h := blockPos.Side(f); w.Block(h) == fire() {
+					w.SetBlock(h, air(), nil)
```
Comment (T14Raptor) [inline server/entity/splashable.go:80]: Same as here

File: server/entity/area_effect_cloud.go (hunk 2 above)
Comment (JustTalDevelops) [reply server/entity/area_effect_cloud.go:96]: I think it's better now? I also removed the locks/unlocks for each entity.

File: server/entity/splashable.go (hunk 3 above)
Comment (JustTalDevelops) [reply server/entity/splashable.go]: Changed.

File: server/entity/splashable.go (hunk 4 above)
Comment (JustTalDevelops) [reply server/entity/splashable.go]: Changed.

File: server/entity/area_effect_cloud.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	transform
+}
+
+// NewDefaultAreaEffectCloud ...
+func NewDefaultAreaEffectCloud(pos mgl64.Vec3, t potion.Potion) *AreaEffectCloud {
+	return NewAreaEffectCloud(pos, t, time.Second*30, time.Second*2, 0, 3.0, -0.5, -0.005)
+}
+
+// NewAreaEffectCloud ...
+func NewAreaEffectCloud(pos mgl64.Vec3, t potion.Potion, duration, reapplicationDelay, durationOnUse time.Duration, radius, radiusOnUse, radiusGrowth float64) *AreaEffectCloud {
```
Comment (Sandertv) [inline server/entity/area_effect_cloud.go:40]: I would name `NewDefaultAreaEffectCloud` to `NewAreaEffectCloud` and `NewAreaEffectCloud` to something else, like `NewAreaEffectCloudWith` or something.

**Review by Sandertv — APPROVED** [review]
This looks good to me, thanks for the PR! I've got one small comment for changing two function names. Should be good once resolved and once conflicts are resolved.

File: server/entity/area_effect_cloud.go (hunk 5 above)
Comment (JustTalDevelops) [reply server/entity/area_effect_cloud.go]: Renamed.

**Comment by JustTalDevelops** [comment]
This should be ready to merge now.

**Pull request merged**